client/rest/src/plugins/metadata/metal.js

```diff
@@ -0,0 +1,214 @@
+
+		const scopedMetadataKey = catapult.utils.uint64.fromBytes(new Uint8Array(b.subarray(4, 12)).reverse());
+		const chunkPayload = b.subarray(12);
+
+		return {
+			magic, text, scopedMetadataKey, chunkPayload
+		};
+	},
+
+	/**
+	 * Separate by null if text is present or not.
+	 * @param {object} chunkData chunkPayload and has text or not
+	 * - chunkPayload: Buffer
+	 * - text: boolean
+	 * @returns {object} return chunkPayload and chunkText.
+	 * - chunkPayload: Uint8Array
+	 * - chunkText: Uint8Array
+	 */
+	splitChunkPayloadAndText(chunkData) {
+		if (!chunkData.text) {
+			// No text in the chunk
+			return {
+				chunkPayload: chunkData.chunkPayload,
+				chunkText: undefined
+			};
+		}
+
+		// Extract text section until null char is encountered.
+		const textBytes = [];
+		for (let i = 0; i < chunkData.chunkPayload.length && chunkData.chunkPayload[i]; i++)
+			textBytes.push(chunkData.chunkPayload[i]);
+
+		return {
+			chunkPayload: new Uint8Array(chunkData.chunkPayload.subarray(textBytes.length + 1)),
+			chunkText: new Uint8Array(textBytes)
+		};
+	},
+
+	/**
+	 * Decode binary data from chunks.
+	 * @param {Uint64} fitstKey ScopedMetadataKey of the beginning chunk.
+	 * @param {Array} chunks Chunk containing data to be decoded.
+	 * @returns {Buffer} Decoded binary data.
+	 */
+	decode(fitstKey, chunks) {
+		let scopedMetadataKey = fitstKey;
+		let magic;
+		let decodedPayload = Buffer.from([]);
+		let decodedText = Buffer.from([]);
+		const findChunk = c => 0 === catapult.utils.uint64.compare(c.key, scopedMetadataKey);
+		do {
+			const chunk = chunks.find(findChunk);
+			if (!chunk)
+				throw new Error(`Error: The chunk ${scopedMetadataKey} is missing`);
+
+			const checksum = this.generateMetadataKey(chunk.value);
+			const isValidChecksum = 0 === catapult.utils.uint64.compare(checksum, scopedMetadataKey);
+			if (!isValidChecksum)
+				throw new Error(`Error: The chunk ${scopedMetadataKey} is broken (calculated=${checksum})`);
+
+			const chunkData = this.extractChunk(chunk.value);
+			({ magic, scopedMetadataKey } = chunkData);
+			const { chunkPayload, chunkText } = this.splitChunkPayloadAndText(chunkData);
+			if (chunkPayload.length)
+				decodedPayload = Buffer.concat([decodedPayload, chunkPayload]);
+
+			if (chunkText?.length)
+				decodedText = Buffer.concat([decodedText, chunkText]);
+		} while (0x80 !== (magic & 0x80));
+		return {
+			payload: decodedPayload,
+			text: decodedText.length ? decodedText.toString('utf-8') : undefined
+		};
+	},
+
+	/**
+	 * Get metadata entry by compositehash.
+	 * I can't throw exceptions in db tests, so I made it for unit tests.
+	 * @param {object} metadata Metadata obtained by composite hash.
+	 * @returns {object} metadataEntry
+	 */
+	getMetadataEntryByCompositehash(metadata) {
+		if (0 === metadata.length)
+			throw Error('could not get first chunk, it may mistake the metal ID.');
+		return metadata[0].metadataEntry;
+	}
+};
+
+class MetalSeal {
```

Comment:
do you mean i need to add JSDoc?
```js
/**
 * MetalSeal class
 */
class MetalSeal {
	static SCHEMA = 'seal1';

	static COMPAT = [MetalSeal.SCHEMA];

	/**
	 * @param {number} length size of the file
	 * @param {string} mimeType mime type of the file
	 * @param {string} name file name of the file
	 * @param {string} comment some comment
	 */
	constructor(length, mimeType, name, comment) {
		this.length = length;
		this.mimeType = mimeType;
		this.name = name;
		this.comment = comment;
		this.schema = MetalSeal.SCHEMA;
	}

	/**
	 * Checks if the given value is a MetalSeal head
	 * @param {any} value - The value to check
	 * @returns {boolean} - Returns true if the value is a MetalSeal head, false otherwise
	 */
	static isHead(value) {
		return value !== undefined
		&& MetalSeal.SCHEMA === value[0]
		&& 'number' === typeof (value[1])
		&& ('string' === typeof (value[2]) || !value[2])
		&& ('string' === typeof (value[3]) || !value[3])
		&& ('string' === typeof (value[4]) || !value[4]);
	}

	/**
	 * Converts the MetalSeal object into a JSON string.
	 * The resulting string includes the schema, length, mimeType (if exists), name (if exists), and comment (if exists).
	 * @returns {string} - A JSON string representation of the MetalSeal object.
	 */
	stringify() {
		const result = [
			this.schema,
			this.length
		];

		if (this.mimeType)
			result.push(this.mimeType);
		else if (this.name || this.comment)
			result.push(null);

		if (this.name)
			result.push(this.name);
		else if (this.comment)
			result.push(null);

		if (this.comment)
			result.push(this.comment);

		return JSON.stringify(result);
	}

	/**
	 * Parses a JSON string into a MetalSeal object.
	 * The JSON string should represent an array with the schema, length, mimeType (if exists), name (if exists), and comment (if exists).
	 * @param {string} json - The JSON string to parse.
	 * @returns {MetalSeal} - A new MetalSeal object created from the parsed JSON string.
	 * @throws {Error} - Throws an error if the JSON string is malformed or does not represent a valid MetalSeal object.
	 */
	static parse(json) {
		const parsedObject = JSON.parse(json);
		if (!Array.isArray(parsedObject)
			|| !this.isHead(parsedObject)
			|| !MetalSeal.COMPAT.includes(parsedObject[0])
		)
			throw new Error('Malformed seal JSON.');

		return new MetalSeal(
			parsedObject[1],
			parsedObject[2] ?? undefined,
			parsedObject[3] ?? undefined,
			parsedObject[4] ?? undefined,
			parsedObject[0]
		);
	}
}
```